### DISCUSSION TITLE
How To: Moving Library locations without losing metadata

### DISCUSSION THREAD

**Discussion by taylorsilva** [jellyfin/jellyfin#6924] [Show and tell] [answered]
## Update October 2022

@MMMZZZZ has a python script that can do the migration. Script can be found here: https://github.com/MMMZZZZ/Jellyfin-Migrator

He definitely dug a lot deeper into the db structure than I did. Going with his script is probably the safer method. See his comment here for details on what my guide below misses: https://github.com/jellyfin/jellyfin/discussions/6924#discussioncomment-3943593

---

## Original Guide

This is a guide going over how to move the location of your library files without losing all the metadata you've built it. This might not be the right or best way but it worked for me. This is basically a more detailed version of [this reddit thread](https://www.reddit.com/r/jellyfin/comments/gkgbkv/what_is_the_best_method_for_moving_my_media_files/). Leave a comment if you know a better way!

I did this on an Ubuntu 20.04 running Jellyfin 10.7.7 in Docker.

## The Problem

I had all of my library files split over two external drives. The drives were mounted at:

* `/media/VolumeOne`
* `/media/VolumeTwo`

Each mount contained the folders `Movies` and `TVShows`. Both these drives were almost full and one was showing signs of failing, so I wanted to migrate everything to one larger drive. I mounted this drive at `/media/Volume5TB` and created the folders `Movies` and `TVShows`. I then copied everything over to this new drive.

```bash
# -r means recursive and -v shows what file it's currently copying over
cp -r -v /media/VolumeOne/TVShows/* /media/Volume5TB/TVShows/
cp -r -v /media/VolumeOne/Movies/* /media/Volume5TB/Movies/
cp -r -v /media/VolumeTwo/TVShows/* /media/Volume5TB/TVShows/
cp -r -v /media/VolumeTwo/Movies/* /media/Volume5TB/Movies/
```

So everything is migrated over. Now I have a problem. How do I keep all that metadata I've built up in Jellyfin?

## Migrating Library Media

There are two places where things need to be updated to point to the new drive. The first is in the database and the second is in Jellyfin's config files.

### 1. Make A Backup

I fucked up many times doing this and the only thing that saved me was having a backup. Have this ready in case you mess up and want to reset everything back to its last known working state.

1. Shutdown the Jellyfin server
2. Make the backup: `tar -czf jellyfin-config-backup.tgz /path/to/jellyfin/config`

Now that you made your backup let's start messing around with Jellyfin's data.

### 2. Updating Database

I used [DB Browser for SQLite](https://sqlitebrowser.org/) to connect to the database file and run queries. The file you need to connect to, located in Jellyfin's `data` folder, is `library.db`.

There are two tables that need to be updated: `TypedBaseItems` and `mediastreams` (not actually sure that `mediastreams` needs to be updated. Didn't hurt anything in my case at least).

The queries for both are similar. Both tables have a `Path` column that contain the entire literal path to your media files. So in my case I need to replace all instances of `/media/VolumeOne` and `/media/VolumeTwo` with `/media/Volume5TB`. These are the queries I used:

```sql
-- TypedBaseItems table
-- VolumeOne
UPDATE TypedBaseItems SET
    Path = replace(Path, '/media/VolumeOne/TVShows', '/media/Volume5TB/TVShows')
WHERE
    Path LIKE '/media/VolumeOne/TVShows%';

UPDATE TypedBaseItems SET
    Path = replace(Path, '/media/VolumeOne/Movies', '/media/Volume5TB/Movies')
WHERE
    Path LIKE '/media/VolumeOne/Movies%';

-- VolumeTwo
UPDATE TypedBaseItems SET
    Path = replace(Path, '/media/VolumeTwo/TVShows', '/media/Volume5TB/TVShows')
WHERE
    Path LIKE '/media/VolumeTwo/TVShows%';

UPDATE TypedBaseItems SET
    Path = replace(Path, '/media/VolumeTwo/Movies', '/media/Volume5TB/Movies')
WHERE
    Path LIKE '/media/VolumeTwo/Movies%';

-- mediastreams table
-- VolumeOne
UPDATE mediastreams SET
    Path = replace(Path, '/media/VolumeOne/TVShows', '/media/Volume5TB/TVShows')
WHERE
    Path LIKE '/media/VolumeOne/TVShows%';

UPDATE mediastreams SET
    Path = replace(Path, '/media/VolumeOne/Movies', '/media/Volume5TB/Movies')
WHERE
    Path LIKE '/media/VolumeOne/Movies%';

-- VolumeTwo
UPDATE mediastreams SET
    Path = replace(Path, '/media/VolumeTwo/TVShows', '/media/Volume5TB/TVShows')
WHERE
    Path LIKE '/media/VolumeTwo/TVShows%';

UPDATE mediastreams SET
    Path = replace(Path, '/media/VolumeTwo/Movies', '/media/Volume5TB/Movies')
WHERE
    Path LIKE '/media/VolumeTwo/Movies%';
```

You can verify everything worked by running these queries and seeing zero records returned:
```sql
SELECT COUNT(*) FROM TypedBaseItems
WHERE
Path LIKE '/media/VolumeOne%' OR
Path LIKE '/media/VolumeTwo%'

SELECT COUNT(*) FROM mediastreams
WHERE
Path LIKE '/media/VolumeOne%' OR
Path LIKE '/media/VolumeTwo%'
```

Once you've updated everything in those two tables you can move on to updating Jellyfin's local config files.

### 3. Updating Jellyfin Config Files

The following files need to be updated.

#### `root/default/{Library name}/options.xml`

You'll need to do this for each Jellyfin library you're updating. Near the top of this XML file is a section called `PathInfos`. Mine looked like this:

```xml
  <PathInfos>
    <MediaPathInfo>
      <Path>/media/VolumeOne/Movies</Path>
    </MediaPathInfo>
    <MediaPathInfo>
      <Path>/media/VolumeTwo/Movies</Path>
    </MediaPathInfo>
  </PathInfos>
```
I updated it to this:
```xml
  <PathInfos>
    <MediaPathInfo>
      <Path>/media/Volume5TB/Movies</Path>
    </MediaPathInfo>
  </PathInfos>
```

#### `root/default/{Library name}/{library-name}.mblink`

Each line of this file contains the entire path that's also in the previous XML file. I have no clue why this is needed as it seems redundant with the info in the XML file.

The file looked like this for me:
```
/media/VolumeOne/Movies
/media/VolumeTwo/Movies
```
and I updated it to this:
```
/media/Volume5TB/Movies
```

NOTE: It's VERY important that this file **does not end** with a [newline character](https://en.wikipedia.org/wiki/Newline). If you're using Vi/Vim/Neovim then you must do `set noeol binary` before saving the file. Another option that could work is doing `echo -n /some/path > some.mblink` where `-n` should result in no newline being added. If there is a newline in the file then Jellyfin will not be able to scan your library and the logs will contain an error that alludes to the newline being present; the log line will carry over to the next line. This is some bad file parsing IMO.

Not sure about other shells, but with `zsh` I can verify that there's no Newline by `cat`'ing the file and seeing a percent (`%`) symbol printed at the end:

<img width="432" alt="image" src="https://user-images.githubusercontent.com/4583837/143781725-dd543f24-2fc7-425f-80ef-eff7b5ad9bf6.png">

#### `root/default/{library name}/{library name}.collection`

This file was empty for me so I left it as-is.

#### `data/{collections,playlists}` Folders and XML Files

Both the `collections` and `playlists` folders under `data` will have XML files with paths that need to be updated. At this point it should be pretty straight forward what needs to change so I won't share an example.

### You're Done!

If you want to do one last check for any extra files that need to be updated you can run:
```
ag -l /media/Volume /path/to/jellyfin/config/*
```
And you'll get a list of files that contain the string `/media/Volume`.

You can now start up Jellyfin. Depending on your settings Jellyfin may not scan immediately so feel free to manually start one via the web dashboard.

Watch the log files while it's scanning for any errors. I have a ~2TB library and it took Jellyfin ~30min to scan everything. You'll know everything is good when a scan completes successfully.

Hope this helps folks that end up in the same situatin as me. I wish this guide existed before I started but alas it didn't. Now it does!

### Troubleshooting

If Jellyfin is having trouble while rescanning the library after all the changes you made or you're noticing stuff disappear from your dashboard you should perform the following troubleshooting steps.

1. Review Jellyfin's logs and note down the ID's of any records Jellyfin is returning errors for
2. Find the record for the ID's Jellyfin complained about in `TypedBaseItems` and either a) update the record (depends what Jellyfin complained about, bad path maybe?) or b) delete the record (I had to delete a few duplicate/redundant records since I was merging two volumes into one)
3. Restore from backup and apply your updated db and related files again

I had to repeat this process quite a few times before I had eliminated all bad records.

**Comment by harry71019** [comment]
I'll add in some input here, inside the `TypedBaseItems` there is a column `Images` which points to the image file of the movie/episode/person/etc. This also needs to be changed to use the new path.

**Comment by michaelkrieger** [comment]
I'll add as well that this does move most data, but certainly isn't perfect for consolidating two folders and some history is likely lost.  If you run
```
SELECT Name,Path,HEX(guid) as guid,HEX(ParentId) as parentid,TopParentId,PresentationUniqueKey,UserDataKey FROM typedbaseitems WHERE hex(ParentId) = (SELECT hex(guid) FROM typedbaseitems WHERE type = "MediaBrowser.Controller.Entities.AggregateFolder") ORDER BY Name
```
you'll see that the folders are now duplicated in the database.  Jellyfin, on launch, will log `Removing item, Type: "Folder", Name: "..", Path: "...", Id: ...`, reduce your cache folder size substantially and rescan everything that was once in that folder, finding the new files and adding them to the library anew. The GUIDs it's using are in both the typedbaseitems and ancestorids tables.

I've made a brief attempt at consolidating them and added it to my existing tutorial on it (which uses a bunch of search and replace commands). I still haven't been able to master what I'm missing on consolidating things in the database.  Currently the last step results in things not working when consolidating folders, but likely works fine if renaming folders.  I welcome anyone who knows more than me about this to help fix the final database consolidation so we can have clean silent-to-Jellyfin migrations.

----
**1.** Shutdown Jellyfin

**2.** Move all data across to the new hard drive

**3.** If applicable, change your mapped paths/docker volume maps

**4.** Update paths in on-disk files:
_Make the destination folders different if you're not consolidating_
```
  $ find ./root/default/ -type f -name '*.mblink' -exec sed -i 's%/media/drive1/%/media/pool/%g ; s%/media/drive2/%/media/pool/%g' {} \;
  $ find ./root/default/ -type f -name "options.xml" -exec sed -i 's%/media/drive1/%/media/pool/%g ; s%/media/drive2/%/media/pool/%g' {} \;
  $ find ./data/collections/ -type f -name "collection.xml" -exec sed -i 's%/media/drive1/%/media/pool/%g ; s%/media/drive2/%/media/pool/%g' {} \;
  $ find ./data/playlists/ -type f -name "playlist.xml" -exec sed -i 's%/media/drive1/%/media/pool/%g ; s%/media/drive2/%/media/pool/%g' {} \;
```

**5.** If you consolidated two folders (instead of moved them separately), you'll find that you may have two ./root/default/*/*.mblink files in the same folder pointing to the same place.  In this case I had "Documentaries1.mblink" and "Documentaries.mblink".  Delete one of the duplicates.

**6.** Update paths in the database:
_Make the destination folders different if you're not consolidating_
```
  $ sqlite3 ./data/library.db
    UPDATE mediastreams SET Path = REPLACE(Path, '/media/drive1/', '/media/pool/');
    UPDATE mediastreams SET Path = REPLACE(Path, '/media/drive2/', '/media/pool/');
    UPDATE TypedBaseItems SET Path = REPLACE(Path, '/media/drive1/', '/media/pool/');
    UPDATE TypedBaseItems SET Path = REPLACE(Path, '/media/drive2/', '/media/pool/');
    UPDATE TypedBaseItems SET data = REPLACE(data, '/media/drive1/', '/media/pool/');
    UPDATE TypedBaseItems SET data = REPLACE(data, '/media/drive2/', '/media/pool/');
    UPDATE TypedBaseItems SET Images = REPLACE(Images, '/media/drive1/', '/media/pool/');
    UPDATE TypedBaseItems SET Images = REPLACE(Images, '/media/drive2/', '/media/pool/');
    UPDATE Chapters2 SET ImagePath = REPLACE(ImagePath, '/media/drive1/', '/media/pool/');
    UPDATE Chapters2 SET ImagePath = REPLACE(ImagePath, '/media/drive2/', '/media/pool/');
```

**7.** Update the paths in AggregateFolder entity manually
_Make the destination folders different if you're not consolidating_
```
  $ sqlite3 ./data/library.db
    SELECT data FROM typedbaseitems WHERE type = "MediaBrowser.Controller.Entities.AggregateFolder";
    _// IMPORTANT manually remove the extra folders in the PhysicalLocationsList and fill in the below data._
    UPDATE typedbaseitems SET data='{"VirtualChildren"...false}' WHERE type = "MediaBrowser.Controller.Entities.AggregateFolder";
    .quit

```
**8.** You'll need to manually update the NetworkPath entries in ./root/default/*/options.xml or do so via the interface after startup

**9.** If you're consolidating two folders like I was, you'll want to manually edit ./root/default/*/options.xml and remove duplicate MediaPathInfo entries.

**10.** Consolidate IDs in the database to avoid a full rescan of your database and removal of everything on launch
IMPORTANT: _This step is ONLY NEEDED if you are merging folders and NOT if you are simply renaming an existing folder to a new place.
This also doesn't yet work._

**10A.** View the duplicate parent IDs that we need to consolidate
```
  $ sqlite3 ./data/library.db
    SELECT Name,Path,HEX(guid) as guid,HEX(ParentId) as parentid,TopParentId,PresentationUniqueKey,UserDataKey FROM typedbaseitems WHERE hex(ParentId) = (SELECT hex(guid) FROM typedbaseitems WHERE type = "MediaBrowser.Controller.Entities.AggregateFolder") ORDER BY Name
```

**###WARNING### NOT YET COMPLETED/WORKING/TESTED**
#**10B.** For each pair of library folders, look up the guid (sourceGUID) and TopParentID (sourceTopParentID) and find its counterpart (destGUID) and (destTopParentID).
```
  $ sqlite3 ./data/library.db
    UPDATE TypedBaseItems Set ParentId=x'destGUID', TopParentId=x'destTopParentID' WHERE ParentId = x'sourceGUID';
    # Delete the old unused one
    DELETE from TypedBaseItems WHERE guid = x'sourceGUID';
    DELETE from AncestorIds WHERE itemId = x'sourceGUID';
    # SKIP: You don't seem to need to update ancestorid as it is common for the library and not the folder
    # UPDATE ancestorids SET ancestorsidtext='destGUID' WHERE itemid = (select ancestoridtext FROM ancestorids WHERE ItemId = x'sourceGUID');
```

**Reply by harry71019** [reply]
I'm moving from a windows install to a docker on a NAS, at first following @taylorsilva the migration seems to work, but when jellyfin rescans it just removes the current entries and rebuilds which is a mess. @michaelkrieger I was wondering if you encountered this issue and how did you solve it? Any thoughts is appreciated, thanks!

**Reply by taylorsilva** [reply]
I had that issue many times while trying to do this. From what I recall now the problem was always that I missed some items or I messed up the paths somehow. It always came down to bad db surgery on my part. You should make sure you remove ANY references to the old path that you can find in the db.

**Reply by taylorsilva** [reply]
Another thing I found useful was viewing jellyfin's logs to see what it said it couldn't find or was having trouble reconciling. I would note down the ids it complained about and then either:

1. Review and update that record if any of it needed updating. Then restore from backup and try the whole thing over again

Or

2. Delete the record that Jellyfin complained about. I had to do this for ~4-ish records I think. Then restore from backup and try again.

Looking at Michael's post I think his points are covering those two cases, so probably best to try what he's suggesting.

**Reply by harry71019** [reply]
I think I still haven't figure out what's the issue here. @michaelkrieger, in step 7, do I change that data to `"VirtualChildren"...false` or do I just edit the PhysicalLocations and keep the rest like this
{
    "DateLastSaved": "2020-07-20T06:35:37.9164273Z",
    "ExtraIds": [],
    "Height": 0,
    "IsHD": false,
    "IsRoot": true,
    "IsShortcut": false,
    "LinkedChildren": [],
    "PhysicalLocationsList": [
        "/config/root",
        "/media/Anime",
        "/media/Live Performance",
        "/media/Movies",
        "/media/MV",
        "/media/TV"
    ],
    "RemoteTrailers": [],
    "SupportsExternalTransfer": false,
    "VirtualChildren": [
        {
            "DateLastSaved": "2020-01-04T04:36:31.735964Z",
            "ExtraIds": [],
            "Height": 0,
            "IsHD": false,
            "IsShortcut": false,
            "RemoteTrailers": [],
            "SupportsExternalTransfer": false,
            "Width": 0
        }
    ],
    "Width": 0
}

**Reply by michaelkrieger** [reply]
SQLite doesn't have JSON extensions, so we can't, using SQL alone, simply update the paths.  Copy and paste that whole JSON block into Notepad or your favourite editor, change the PhysicalLocationsList to correctly reflect your new locations, and then take the whole block and update the row with the new list using that SQL command.  JSON is just formatted text string.
```
"PhysicalLocationsList": [
"/config/root",
"/media/Anime",
"/media/Live Performance",
"/media/Movies",
"/media/MV",
"/media/TV"
],
```

If you do everything right, there is no reason why the database won't just see all is correct and not do any rescanning or adding if we get all of the database and file updates.  Then again, I was _consolidating_ two folders (turning /media/disk0 and /media/disk0 into /media/pool), so if you're just updating the paths, you probably don't need step 10 at all and in theory everything should work.

**Reply by harry71019** [reply]
@michaelkrieger  Thank you for the quick reply and I'll try that. I'm migrating from windows to synology docker so I'm mostly just changing the paths and the media folder structure is the same. I can move the database and browse/watch it without any problem until I rescan something. I'm wondering if there's a place where Jellyfin logs what stuff is already in the database and what's not?

PS. Sorry if this question seems dumb, I used the command in wsl to swap the data but I was wondering what's the difference from just editing that row in a program like DB browser for SQLite?

**Comment by MMMZZZZ** [comment, marked as answer]
The suggestions here are not wrong but not complete either (which is why it doesn't survive a library scan). I wrote a script that - at the time of writing - successfully migrated the Windows installations of two jellyfin users to Docker without any data loss. While I'd consider the various comments in the script as the better documentation, I thought I'd put a quick overview here since this topic might be linked elsewhere. 

The reason you can't really do it by hand is because the IDs jellyfin uses throughout its database and file paths which uniquely identify every element of the library are in fact derived from the file paths of said elements. Basically a hash of it. So you not only need to update all the file paths but also all the IDs of all elements and all their occurences. These also occur in jellyfins internal metadata file paths, so those must be reorganized, too. 
The second reason is (I think; not 100% sure on this one) that jellyfin stores all the file creation and modification dates, so my script updates them, too. A mismatch of these likely causes a scan event which may not hurt (not sure if it would replace anything or not), but it's not necessary either. 

Finally, here's the link. I don't claim it to be the cleanest piece of software but it's fairly well documented and commented I think. 
https://github.com/MMMZZZZ/Jellyfin-Migrator

@taylorsilva maybe you could update your post...?

Kind regards,
Max

**Reply by taylorsilva** [reply]
Great info! I'll update my post with that. Did not know at all about the file hash thing, which explains a bunch of the "weird" things that I didn't understand happening 😅

**Reply by KeshavKhanth** [reply]
Hey there,
For my media server. I recently changed my source path as well as file name structure. I also create a py script to update filename in library.db. I just has some doubt this https://github.com/MMMZZZZ/Jellyfin-Migrator will auto generate item id based on new full path. can you explain the process in details like how it's generating id.

**Reply by Dobeemixer** [reply]
The script didn’t seem to work for me. I tried to test with one path go from "E:/Documents and etc/TV+Movies/Movies": "O:/Media/Movies"
And mapped all the database files correctly, then removed the entries for fs_path_replacement except the specified ones, finally put the source roots and target roots in different folders. Once the script was all done I put the new library.db into the old location to see if the new paths transferred, and they didn’t. My movies folder shows nothing and the rest of my library crashes when I try to load a video. Not sure what I’m doing wrong here?


EDIT: here's the snippit of edited code. If i read everything right, this was all I need to do for simply migrating the databases references of one folder to another. All the files are in the new folder as well, so I'm a bit confused. (And I’m also planning on keeping my server files in the same areas as before)
...
```
# Please specify a log file. The script is rather verbose and important
# errors might get lost in the process. You should definitely check the
# log file after running the script to see if there are any warnings or
# other important messages! Use f.ex. notepad++ (npp) to quickly
# highlight and remove bunches of uninteresting log messages:
#   * Open log file in npp
#   * Go to "Search -> Mark... (CTRL + M)"
#   * Tick "Bookmark Line"
#   * Search for strings that (only!) occur in the lines you want to
#     remove. All those lines should get a marking next to the line number.
#   * Go to "Search -> Bookmark -> Remove Bookmarked Lines"
#   * Repeat as needed
# Text encoding is UTF-8 (in npp selectable under "Encoding -> UTF-8")
log_file = "F:/jf-migrator.log"


# These paths will be processed in the order they're listed here.
# This can be very important! F.ex. if specific subfolders go to a different
# place than stuff in the root dir of a given path, the subfolders must be
# processed first. Otherwise, they'll be moved to the same place as the other
# stuff in the root folder.
# Note: all the strings below will be converted to Path objects, so it doesn't
# matter whether you write / or \\ or include a trailing / . After the path
# replacement it will be converted back to a string with slashes as specified
# by target_path_slash.
# Note2: The AppDataPath and MetadataPath entries are only there to make sure
# the script recognizes them as actual paths. This is necessary to adjust
# the (back)slashes as specified. This can only be done on "known" paths
# because (back)slashes occur in other strings, too, where they must not be
# changed.
path_replacements = {
    # Self-explanatory, I guess. "\\" if migrating *to* Windows, "/" else.
    "target_path_slash": "/",
    # Paths to your libraries
    "E:/Documents and etc/TV+Movies/Movies": "O:/Media/Movies",
    "E:/Documents and etc/TV+Movies/Lampshade": "E:/Documents and etc/TV+Movies/Lampshade",
    # Paths to the different parts of the jellyfin database. Determine these
    # by comparing your existing installation with the paths in your new
    # installation.
    "C:/ProgramData/Jellyfin/Server/config": "C:/ProgramData/Jellyfin/Server/config2",
    "C:/ProgramData/Jellyfin/Server/cache": "C:/ProgramData/Jellyfin/Server/cache2",
    "C:/ProgramData/Jellyfin/Server/log": "C:/ProgramData/Jellyfin/Server/log2",
    "C:/ProgramData/Jellyfin/Server": "/config/data", # everything else: metadata, plugins, ...
    "C:/ProgramData/Jellyfin/Server/transcodes": "C:/ProgramData/Jellyfin/Server/transcodes2",
    "C:/Program Files/Jellyfin/Server/ffmpeg.exe": "C:/ProgramData/Jellyfin/Server/ffmpeg.exe2",
    "%MetadataPath%": "%MetadataPath%",
    "%AppDataPath%": "%AppDataPath%",
}


# This additional replacement dict is required to convert from the paths docker
# shows to jellyfin back to the actual file system paths to figure out where
# the files shall be copied. If relative paths are provided, the replacements
# are done relative to target_root.
#
# Even if you're not using docker or not using path mapping with docker,
# you probably do need to add some entries for accessing the media files
# and appdata/metadata files. This is because the script must read all the
# file creation and modification dates *as seen by jellyfin*.
# In that case and if you're sure that this list is 100% correct,
# *and only then* you can set "log_no_warnings" to True. Otherwise your logs
# will be flooded with warnings that it couldn't find an entry to modify the
# paths (which in that case would be fine because no modifications are needed).
#
# If you actually don't need any of this (f.ex. running the script in the
# same environment as jellyfin), remove all entries except for
#   * "log_no_warnings" (again, can be set to true if you're sure)
#   * "target_path_slash"
#   * %AppDataPath%
#   * %MetadataPath%.
fs_path_replacements = {
    "log_no_warnings": False,
    "target_path_slash": "/",
    "%AppDataPath%": "/data/data",
    "%MetadataPath%": "/data/metadata",
}


# Original root only needs to be filled if you're using auto target paths _and_
# if your source dir doesn't match the source paths specified above in
# path_replacements.
# auto target will first replace source_root with original_root in a given path
# and then do the replacement according to the path_replacements dict.
# This is required if you copied your jellyfin DB to another location and then
# start processing it with this script.
original_root = Path("C:/ProgramData/Jellyfin/Server")
source_root = Path("C:/ProgramData/Jellyfin/Server")
target_root = Path("E:/Jellyfin-dummy")

```

**Reply by crawdaddydoo92** [reply]
Will your script be helpful if I'm just moving my library to a bigger hard drive?  Also, since I may be doing so in the future, will it work to move from one Linux machine to another on the same network but with a different IP Address?

**Comment by solbu** [comment]
> I had all of my library files split over two external drives. The drives were mounted at:
> 
>     * `/media/VolumeOne`
> 
>     * `/media/VolumeTwo`

>  I mounted this drive at /media/Volume5TB

What I did on my Linux server was to copy the files to the new drive, but keep the folder structure.
So using your paths as the example I would copy  `/media/VolumeOne` to the new drive at `/media/Volume5TB/VolumeOne`
and  `/media/VolumeTwo` to `/media/Volume5TB/VolumeTwo`.

When the copying is complete, I unmount the old drives, delete the old folders which are now empty (because the drives are unmounted) in /media and create symlinks in `/media/` pointing to the new location.
`ln -s /media/Volume5TB/VolumeOne /media/VolumeOne`
`ln -s /media/Volume5TB/VolumeTwo /media/VolumeTwo`

Now jellyfin should work as before. No reconfiguring needed.

**Comment by oatsvine** [comment]
See issue #9922, would it make sense for Jellyfin to implement a "move" action which handles the metadata update concerns while delegating the file system operations to the user?

**Comment by adiberk** [comment]
I don't know if any new solutions came out.  But in my case I would love to be able to simply move one  of my tv shows to another folder on another hard drive. Am I correct that as of now, this is not simple to do? (Ie. I add the new hard drive path to the library, and then move the tv show to it)

I tested it out and jellyfin just removes the show and doesn't register it in the new path.

**Reply by harry71019** [reply]
Unfortunately I think it's not simple. When I did it back in 10.7 I did the following

- First I made a backup of the database
- Since the index of the database is based on path to files, I rescanned the database
- Now all the metadata should be a mess, so I then wrote a python script to match the old metadata (so basically the path for each movie or episode) to the new one based on the title names, and then move all the images to the correct location.

It was really a pain...

**Reply by adiberk** [reply]
That definitely sounds painful.  I have great respect the jellyfin team and know how hard it is to maintain the project. I hope they look deeper into how we can make this process more seamless.  It seems like a very common scenario, especially when people want to start splitting library into multiple drives.

What worked for me is I copied folder to the new drive and renamed it.  Then from jellyfin, deleted the old one. Then renamed and rescanned and the new one showed up with same watch history etc just new path migrated

**Reply by ghstbrry** [reply]
@harry71019

were you storing your metadata in .nfo files or did you update the metadata paths in the db?

do you happen to still have that script around? sounds very similar to how i'd go about solving the problem myself & could be a great jumping off point.

i'm currently trying to think through the solution. tired of losing metadata any time i decide on a slightly altered path scheme.

**Reply by harry71019** [reply]
@stskvi I tried the nfo route back in 10.7. Unless there has been updates to how Jellyfin gets metadata from .nfos I would not suggest this route. If you happen to store all you metadata with the media files in .nfos maybe worth a try. I can try to dig up the script this weekend and see if it still works, but I think you could also check out the scripts posted by others above.

Like I said, I think the best route is dealing with the library.db and fixing all the sql stuff and moving all the images to the correct place.

**Reply by myownbiggestfan** [reply]
I’m a complete layman when it comes to this stuff. I just lost a bunch of metadata because I had no idea moving my library would wreck it. I got here by googling for solutions but I’m not going to dig through this forum to find the correct scripts, I don’t even know how to run them and I’d be likely to break everything trying to use them. This isn’t me complaining, just explaining why I’ve come here to suggest something that’s is probably a bad idea for some reason I’m not aware of. 

Why not add the functionality of moving a library to Jellyfin itself? If it is possible to run a script that fixes it, it seems that it would be possible to run a script within Jellyfin that just updates the metadata and paths as it moves the files. 

At the very least could a warning be included in the dashboard interface? 

As i explained, I’m no programmer so forgive me if my first suggestion is out of the question or harder than it seems. I stand by the warning suggestion though. That can’t be hard to add.

**Comment by Meru** [comment]
This issue is pretty old and most of the things described here were working just like that, but it seems not in the recent versions of jellyfin server.
Using latest release version i was able to successfully movemetadata to a different path on a different drive by simply replacing old path strings to the new ones. Absolutely zero issues. Everything works. Probably some changes contributed to that.

My question is the following though. If jellyfin has metadata path, why not just store metadata as a relative path to it? Wouldnt it allow easy migration by simply selecting new metadata path?

**Reply by maranarasauce** [reply]
I'm on 10.11.5, where did you replace the path strings? Library.db? I'm migrating from a server from Windows to Linux and need to keep all my metadata intact.

Edit: I figured it out! This is what worked for me in order to move my server from Windows to Ubuntu/Docker Compose. Here's personal notes I took as I went along, probably not accurate but what I took away as a first-time Ubuntu/Docker user.
- Create Jellyfin server backup
- In the backup, in the following files, replace "`$OLD_MEDIA_PATH`" with "`$NEW_MEDIA_PATH`" and any occurence of "\\" with "/"
(This is because the way Docker Compose works is using a kind of sym-link to the hard drive and because Ubuntu doesn't use backslashes respectively)
The backup files to replace:
    Root/`$Library`/options.xml
    Database/BaseItemImageInfos.json
    Database/BaseItems.json
    Database/MediaStreamInfos.json

- Restore Jellyfin server backup on Docker server
- Go to Transcoding tab in the admin dashboard and update the transcoding path

**Reply by Meru** [reply]
> I'm on 10.11.5, where did you replace the path strings? Library.db? I'm migrating from a server from Windows to Linux and need to keep all my metadata intact.

I did it on 10.11.6. You need to change paths in jellyfin.db
Backup feature is broken right now so I wouldnt touch that at least until new version

**Reply by crawdaddydoo92** [reply]
I've been backing up the files containing Metadata to my RAID pool. Assuming the file system is compatible with both OS, and that they don't want to change to a different file system, couldn't they just manually backup the files and move to the new server. Not with the built-in tool in jellyfin but with something else, or even an old fashioned copy and paste?

**Reply by Meru** [reply]
> I've been backing up the files containing Metadata to my RAID pool. Assuming the file system is compatible with both OS, and that they don't want to change to a different file system, couldn't they just manually backup the files and move to the new server. Not with the built-in tool in jellyfin but with something else, or even an old fashioned copy and paste?

You can't just copy previously manually archived files and expect it to work. Jellyfin database is Not portable. The only way it would work is if in new place ALL the file paths will be Exactly as they were before. Including all folders, subfolders, drive letters. If at least one symbol changed somewhere, it is an issue.

**Reply by crawdaddydoo92** [reply]
Making them the same isn't that hard. At least it's doable.

**Reply by maranarasauce** [reply]
> I did it on 10.11.6. You need to change paths in jellyfin.db Backup feature is broken right now so I wouldnt touch that at least until new version

Oh, that's interesting! During this migration, I've been completely relying on the Backup feature and haven't seen any issues. Is this issue only seen on 10.11.6 or should I be having issues as well?

**Reply by Meru** [reply]
> > I did it on 10.11.6. You need to change paths in jellyfin.db Backup feature is broken right now so I wouldnt touch that at least until new version
> 
> Oh, that's interesting! During this migration, I've been completely relying on the Backup feature and haven't seen any issues. Is this issue only seen on 10.11.6 or should I be having issues as well?

This issue should be in all versions that support backup at all. But I haven't followed the bug (or bugs) closely so I am not sure what builds platforms or their combinations are affected. All I know is that me and some other people reported issues and someone made a fix. So right now I am simply waiting for the new version to see how it performs before I may report in things that are already fixed but not released yet.

**Comment by SpaceAgeHero** [comment]
Hey guys,

I'd also like to update my library paths, admittedly rather for cosmetic reasons. It's just the beginning of the paths, rest of the actual structures, where the media is sorted remains the same.

1. Stopped Jellyfin-Server (10.11.11)
2. Made sure to update all occurences of `/old/path` in 
`/var/lib/jellyfin/data/jellyfin.db` (in BaseItems Table using SQL queries with DB Browser for SQ Lite)
`/var/lib/jellyfin/root/default/*/*.mblink`
`/var/lib/jellyfin/root/default/*/options.xml`
3. Updated all media storage mount points
4. Started Jellyfin-Server

From the first impression, everything turned out well. Paths looked right in UI, Media playback worked fine.

However ... then, when I performed a library scan, Jellyfin began to remove and re-add every item, pulling default cover art.
(Playback progress still seemed fine.)

But the worst part is, it even began to re-generate chapter preview images and trickplay.

Is there anything I missed? Any way to avoid getting everything re-added?
